pkg/cloudprovider/provider/openstack/provider.go

```diff
@@ -433,36 +409,63 @@ func (p *provider) Create(machine *v1alpha1.Machine, _ cloud.MachineUpdater, use
 		return nil, osErrorToTerminalError(err, "failed to create server")
 	}
 
-	if ip != nil {
-		// if we want to assign a floating ip to the instance, we have to wait until it is running
-		// otherwise the instance has no port in the desired network
-		instanceIsReady := func() (bool, error) {
-			currentServer, err := osservers.Get(computeClient, server.ID).Extract()
-			if err != nil {
-				tErr := osErrorToTerminalError(err, fmt.Sprintf("failed to get current instance %s", server.ID))
-				if isTerminalErr, _, _ := cloudprovidererrors.IsTerminalError(tErr); isTerminalErr {
-					return true, tErr
-				}
-				// Only log the error but don't exit. in case of a network failure we want to retry
-				glog.V(2).Infof("failed to get current instance %s: %v", server.ID, err)
-				return false, nil
-			}
-			if currentServer.Status == "ACTIVE" {
-				return true, nil
+	if err := waitUntilInstanceIsActive(computeClient, server.ID); err != nil {
+		defer deleteInstanceDueToFatalLogged(computeClient, server.ID)
+		return nil, fmt.Errorf("instance %s became not active: %v", server.ID, err)
+	}
+
+	// Find a free FloatingIP or allocate a new one
+	if c.FloatingIPPool != "" {
+		if err := assignFloatingIPToInstance(client, server.ID, c.FloatingIPPool, c.Region, network); err != nil {
+			defer deleteInstanceDueToFatalLogged(computeClient, server.ID)
```

Comment:
What about instead making the `Create` of the openstack provider idempotent by calling `Get` before attempting to create the instance, then do the FIP stuff? I believe this is what we do in Azure and feels much more elegant to me

Reply:
The controller always executes a `Get` before doing the `Create`. Only if the `Get` cannot be satisfied, a `Create` gets executed.

#### Azure does 2 relevant requests:
- Create the NIC - A `Get` would **not** be satisfied in case we fail here
- Create the VM with the NIC - A `Get` would **not** be satisfied in case we fail here

The `Get` with Azure gets satisfied by finding a VM which has the Machine-UID as a tag.
Therefore only if the last request succeeds, the `Get` would be satisfied.

#### In OpenStack, the flow is a bit different.
- Create the VM -  A `Get` would **not** be satisfied in case we fail here
- Get a FloatingIP -  A `Get` **would** be satisfied in case we fail here
- Assign the Floating IP - A `Get` **would** be satisfied in case we fail here

The `Get` with OpenStack gets satisfied by finding a VM which has the Machine-UID in the VM metadata.
Therefore it would get satisfied as soon as we have been able to create the VM.


A more solid solution would probably be to introduce some kind of `EnsureVMExists` - which replaces the `Get` and `Create` function. Though this has the potential to increase the number of API calls, done to the cloud-provider, to increase by quite a bit.

Reply:
Why not just extend the `Get` of the Openstack provider to not succeed in case a FIP pool was specified but the instance does not have an FIP associated?

Reply:
It would be a special and non expectable implementation of a `Get` operation - as the `Get` would now also need to verify if the instance at the cloud-provider is actually according to spec.
It feels like a very broken mix of a `Get/Create` & `Ensure` approach.

If we want to go that way, we should do it proper and not partially & weird.

Reply:
Okay I see. The reason I am nitpicking on this so much is that I very much dislike that we recreate an instance just because FIP allocation & assignment failed due to concurrency issues.

Maybe we can create an issue to replace the existing `GET` & `Create` with an `Ensure` in the interface? Implementation should be quite trivial.

Reply:
I created https://github.com/kubermatic/machine-controller/issues/318
Though i'm not entirely sure if its such an easy implementation